lib/Twig/Environment.php

```diff
@@ -803,18 +814,28 @@ public function getExtension($name)
      */
     public function addExtension(Twig_ExtensionInterface $extension)
     {
-        $name = $extension->getName();
+        $name = get_class($extension);
+
+        $m = new ReflectionMethod($extension, 'getName');
+        $legacyName = null;
+        // we don't trigger the deprecation notice for the internal extensions
+        // as people might have extended them (even if this is deprecated and not recommended anymore)
+        if ('Twig_Extension' !== $m->getDeclaringClass()->getName() && 'Twig_Extension_' !== substr($name, 0, 15) && 'Mock_Twig_Extension' !== substr($name, 0, 19)) {
+            @trigger_error(sprintf('Defining the getName() method in the "%s" extension is deprecated since 1.26 and will be removed in 2.0. It\'s not used anymore internally.', $name), E_USER_DEPRECATED);
+            $legacyName = $extension->getName();
+        }
 
         if ($this->extensionInitialized) {
             throw new LogicException(sprintf('Unable to register extension "%s" as extensions have already been initialized.', $name));
         }
 
-        if (isset($this->extensions[$name])) {
-            @trigger_error(sprintf('The possibility to register the same extension twice ("%s") is deprecated since version 1.23 and will be removed in Twig 2.0. Use proper PHP inheritance instead.', $name), E_USER_DEPRECATED);
+        if (null !== $legacyName && isset($this->legacyExtensionNames[$legacyName])) {
```

Comment:
And for a reused legacy name, shouldn't this replace the existing extension ? This will not be the case currently, as it does not unregister it (it only unregisters the legacy alias).